### DISCUSSION TITLE
Missing Add Button in Plugin - any ideas why?

### DISCUSSION THREAD

**Discussion by geckigek** [netbox-community/netbox#15314] [Help Wanted!]
Hi there,

I am trying to learn the development of netbox plugins. I am currently trying to adapt the tutorial on to my own plugin. But I am stuck on a problem at step 5: views. At the end of step 5 you should see a add button on the top right to add new objects. But in my case there is no add button. Is there maybe a chance that someone could have a look at my code and maybe tell me what my error is?

Here is my code:

**__init__.py.:**

```python
from extras.plugins import PluginConfig


class VIPConfig(PluginConfig):
    name = "vip_plugin"
    verbose_name = "VIP"
    description = "Manage data structure of VideoIPath"
    version = "0.1"
    base_url = "vip"


config = VIPConfig

``` 

**setup.py:**

```python
from setuptools import find_packages, setup

setup(
    name="vip-plugin",
    version="0.1",
    description="Manage data structure of VideoIPath",
    install_requires=[],
    packages=find_packages(),
    include_package_data=True,
    package_data={"vip_plugin": ["templates/vip_plugin/*.html"]},
    zip_safe=False,
)

``` 

**models.py:**

```python
from django.db import models
from django.urls import reverse
from netbox.models import NetBoxModel


class VideoIPathDevicetype(NetBoxModel):

    name = models.CharField(max_length=100)

    devicetype = models.ForeignKey(
        to="dcim.DeviceType", on_delete=models.CASCADE, blank=False
    )
    extradata = models.JSONField(blank=True, null=True)

    comments = models.TextField(blank=True)

    class Meta:
        ordering = ("name",)

    def __str__(self):
        return self.name

    def get_absolute_url(self):
        return reverse("plugins:vip_plugin:vipdevtype", args=[self.pk])

``` 

**tables.py:**

```python
import django_tables2 as tables

from netbox.tables import NetBoxTable, ChoiceFieldColumn
from .models import VideoIPathDevicetype


class VideoIPathDevicetypeTable(NetBoxTable):

    name = tables.Column(linkify=True)

    class Meta(NetBoxTable.Meta):
        model = VideoIPathDevicetype
        fields = ("pk", "id", "name", "devicetype", "comments")
        default_columns = ("name", "devicetype")

``` 

**forms.py:**
```python
from netbox.forms import NetBoxModelForm
from .models import VideoIPathDevicetype
from utilities.forms.fields import CommentField, DynamicModelChoiceField
from dcim.models import DeviceType


class VideoIPathDevicetypeForm(NetBoxModelForm):

    comments = CommentField()

    devicetype = DynamicModelChoiceField(queryset=VideoIPathDevicetype.objects.all())

    class Meta:
        model = VideoIPathDevicetype
        fields = ("name", "devicetype", "comments", "tags")
``` 

**views.py:**

```python
from django.urls import reverse_lazy
from netbox.views import generic
from . import forms, models, tables


class VideoIPathDevicetypeView(generic.ObjectView):
    queryset = models.VideoIPathDevicetype.objects.all()


class VideoIPathDevicetypeListView(generic.ObjectListView):
    queryset = models.VideoIPathDevicetype.objects.all()
    table = tables.VideoIPathDevicetypeTable


class VideoIPathDevicetypeEditView(generic.ObjectEditView):
    queryset = models.VideoIPathDevicetype.objects.all()
    form = forms.VideoIPathDevicetypeForm


class VideoIPathDevicetypeDeleteView(generic.ObjectDeleteView):
    queryset = models.VideoIPathDevicetype.objects.all()

``` 

**urls.py:**

```python
from django.urls import path
from . import models, views

from netbox.views.generic import ObjectChangeLogView

urlpatterns = (
    path("vipdevtype/", views.VideoIPathDevicetypeListView.as_view(), name="vip_list"),
    path(
        "vipdevtype/add/", views.VideoIPathDevicetypeEditView.as_view(), name="vip_add"
    ),
    path("vipdevtype/<int:pk>/", views.VideoIPathDevicetypeView.as_view(), name="vip"),
    path(
        "vipdevtype/<int:pk>/edit/",
        views.VideoIPathDevicetypeEditView.as_view(),
        name="vip_edit",
    ),
    path(
        "vipdevtype/<int:pk>/delete/",
        views.VideoIPathDevicetypeDeleteView.as_view(),
        name="vip_delete",
    ),
    path(
        "vipdevtype/<int:pk>/changelog/",
        ObjectChangeLogView.as_view(),
        name="vip_changelog",
        kwargs={"model": models.VideoIPathDevicetype},
    ),
)

```

**Comment by geckigek** [comment]
![Missing-Add-Button-Top-Right](https://github.com/netbox-community/netbox/assets/86916063/91bb3803-cada-4c77-8b5f-0053b332406d)

**Comment by mvrp21** [comment]
A bit late, but if anyone ends up with this problem: The issue arises because of the expected name for the urls.

Basically, NetBox can't find the actual `vip_edit` view. And, assuming you didn't really want to make one, NetBox simply doesn't show the add button which is essentially a link to it.


## Detecting the error

One of the ways you can detect this error is by proceeding with the tutorial anyways into the API part and creating a model. You can also do this using the `manage.py nbshell` command, without the API part:

```python
vip = VideoIPathDevicetype(name='whatever', devicetype=DeviceType.objects.all()[0])
vip.save()
```

Then, when you go back to the `https://<your-instance>/plugins/vip/vipdevtype/` URL you'll get the following Exception:

```
NoReverseMatch at /plugins/vip/vipdevtype/

Reverse for 'vipdevtype' not found. 'vipdevtype' is not a valid view function or pattern name.
```

Looking a bit into it, it seems Django looks for a URL with name `vipdevtype`, however you named your URL `vip`. Changing the `reverse()` method for your model like this:

```python
def get_absolute_url(self):
    return reverse("plugins:vip_plugin:vip", args=[self.pk])
```

Fixes that, but results in **another** error:

```
Reverse for 'videoipathdevicetype_edit' not found. 'videoipathdevicetype_edit' is not a valid view function or pattern name.
```


## Let's get more in-depth here

Looking at the traceback, we eventually discover that the reverse being called is the function `get_viewname()` in the `netbox/tables/columns.py` file. [It even seems to be the same function today](https://github.com/netbox-community/netbox/blob/main/netbox/utilities/views.py#L256C17-L256C53).

What is the expecteed viewname for a NetBox plugin? `plugins:{app_label}:{model_name}`. Key here is the `model_name` part.

If you look closely you'll see that in the tutorial all url names are actually the model name in lowercase. Which is the only real difference between this code and the tutorial's code.


The plugin tutorial does say:

> "However, it is recommended to retain the naming scheme shown, as several NetBox features rely on it."

Which is honestly a bit vague.

### Wait a second

> Why did the first url work? (After modifying the `get_absolute_url()` method)

Again, looking at the traceback: The actual model's method is called (`return context.get_absolute_url()`). While in the "action" view's case it calls `url = reverse(get_viewname(model, action), kwargs={'pk': record.pk})` which assumes your url names to be the model's name.


In my opinion this is non-trivial behavior, but it works if you figure it out.

Should I maybe make a PR better specifying what the "naming scheme" even is for the urls, and which "several NetBox features" rely on it? Probably.


## TL;DR

Just set the URL names to be the model name in lowercase. Even if it can be different somehow, why bother?

Update your model's `get_absolute_url` to:

```python
def get_absolute_url(self):
    return reverse("plugins:vip_plugin:videoipathdevicetype", args=[self.pk])
```

And urls to:

```python
urlpatterns = (
    path("vipdevtype/", views.VideoIPathDevicetypeListView.as_view(), name="videoipathdevicetype_list"),
    path(
        "vipdevtype/add/", views.VideoIPathDevicetypeEditView.as_view(), name="videoipathdevicetype_add"
    ),
    path("vipdevtype/<int:pk>/", views.VideoIPathDevicetypeView.as_view(), name="videoipathdevicetype"),
    path(
        "vipdevtype/<int:pk>/edit/",
        views.VideoIPathDevicetypeEditView.as_view(),
        name="videoipathdevicetype_edit",
    ),
    path(
        "vipdevtype/<int:pk>/delete/",
        views.VideoIPathDevicetypeDeleteView.as_view(),
        name="videoipathdevicetype_delete",
    ),
    path(
        "vipdevtype/<int:pk>/changelog/",
        ObjectChangeLogView.as_view(),
        name="videoipathdevicetype_changelog",
        kwargs={"model": models.VideoIPathDevicetype},
    ),
)
```

(Better yet might be to use the actual model name as a variable and not even type it, but this works)

And voilà:

![voila](https://github.com/user-attachments/assets/882ca2af-62db-4b65-82ce-6a240c664ffa)
